### ISSUE TITLE
Microsoft.Graph.Authentication Module v2.0.0 cannot be imported in Azure Automation

### ISSUE THREAD

**Issue by PowershellNinja** [microsoftgraph/msgraph-sdk-powershell#2133]
**Describe the bug**
Import vom Module Microsoft.Graph.Authentication Module v2.0.0 fails in Azure Automation with the error message
```Powershell
Error importing the module Microsoft.Graph.Authentication. Import failed with the following error: 
Orchestrator.Shared.AsyncModuleImport.ModuleImportException: While importing the module, an error occurred while processing 
the module content. Internal error message: Exception has been thrown by the target of an invocation.. 
at Orchestrator.Activities.SetModuleVersion.ExecuteInternal(CodeActivityContext context, String moduleName, 
ModuleLanguage moduleLanguage, String pythonModuleVersion, String modulePath, Guid accountId, Guid moduleVersionId, 
Int64 moduleContentByteSize, String storageUri, Int32 moduleVersion) at 
Orchestrator.Activities.SetModuleVersion.Execute(CodeActivityContext context) at 
System.Activities.CodeActivity.InternalExecute(ActivityInstance instance, ActivityExecutor executor, 
BookmarkManager bookmarkManager) at 
System.Activities.Runtime.ActivityExecutor.ExecuteActivityWorkItem.ExecuteBody(ActivityExecutor executor, 
BookmarkManager bookmarkManager, Location resultLocation)
```

**To Reproduce**
1. Create an Azure Automation Account
2. Go to Modules
3. Go to Add a Module
4. Browse Module Microsoft.Graph.Authentication from Gallery and Import
5. Check Error Message after Module Import shows "Failed"

**Expected behavior**
I expect the module to import. It worked without any issues for module Version 1.28.0.

**Debug Output**
None

**Module Version**
Microsoft.Graph.Authentication v2.0.0

**Environment Data**
None

**Screenshots**
![image](https://github.com/microsoftgraph/msgraph-sdk-powershell/assets/12686296/b76490f7-809b-4bed-bac4-4d0c2efa4583)

**Additional context**
Workaround: Directly go to https://www.powershellgallery.com/packages/Microsoft.Graph.Authentication/1.28.0, choose "Deploy to Azure Automation" on the "Azure Automation" Tab and use v1.28.0 for now.
![image](https://github.com/microsoftgraph/msgraph-sdk-powershell/assets/12686296/d55640ac-ab69-4d23-8425-056ee6cc15e4)

**Comment by peombwa** [comment]
Thanks for bringing this to our attention.

I'm able to install v2.0.0 just fine in an Automation Account (PowerShell 5.1 and 7.2) using the repro steps above:
![image](https://github.com/microsoftgraph/msgraph-sdk-powershell/assets/7061532/1319109c-06c8-4d8b-9bb0-2818a867e400)

The stack trace suggests that this is an installation error in Azure Automation Account when installing the module. Could you retry importing the module in the Automation Account assuming this was a transient error with Azure Automation Account?

If the problem persists, I recommend that you create a support ticket or question with the Azure Automation Account team. See https://learn.microsoft.com/azure/automation/troubleshoot/automation-account.

**Comment by PowershellNinja** [comment]
The installation/upgrade still fails on this particular automation account:
![image](https://github.com/microsoftgraph/msgraph-sdk-powershell/assets/12686296/0506f996-285c-41e6-8430-b6e1eea25171)
I was able to successfully import the module with version 2.0.0 on a newly created Automation Account, so this is probably an issue specific to that one Automation Account:

![image](https://github.com/microsoftgraph/msgraph-sdk-powershell/assets/12686296/1948f53d-c810-4df6-8efe-873d9bf86bfb)

 I'll thus create a support ticket and close this issue, thanks for your assistance @peombwa!

**Comment by bemokraig** [comment]
We are seeing this same problem.  Fails on existing and fails on new AAs.

**Comment by Roladin** [comment]
Also having this issue. Microsoft.Graph.Authentication 2.0.0 stable release fails to import when importing from PS gallery or saving module to zip and uploading manually.

Importing 2.0.0 rc3 from the PS gallery works without issue.

**Comment by PowershellNinja** [comment]
Reopening as more people seem to have this issue

**Comment by A-J-A** [comment]
I'm also having the same issue

**Comment by peombwa** [comment]
Could you try adding the module to Azure Automation again to see if the issue is still present? It does appear that there was an issue in Azure Automation that was preventing modules from being imported, which has since been resolved - https://github.com/Azure/azure-powershell/issues/21647#issuecomment-1632892165.

Please open a support ticket or question with the Azure Automation Account Team if the problem persists as the issue appears to be isolated to a few Automation Accounts (AA). See https://learn.microsoft.com/azure/automation/troubleshoot/automation-account. We are able to successfully import the module on new AA accounts.

@PowershellNinja, did you manage to hear back from the Azure Automation support team? I'll also kickstart an internal comms with the team for awareness if the issue is still present.

**Comment by Roladin** [comment]
@peombwa thanks for following this up. 
I am able to successfully import the latest Microsoft.Graph.Authentication module from PS gallery using runtime version 7.2.
Trying to import with runtime version 5.1 or 7.1 still fails.

**Comment by LOldfield** [comment]
Likewise running into this with 5.1, 7.1
Can't import via the portal or via powershell

**Comment by bemokraig** [comment]
Failed for me today, 5.1.  No change.

**Comment by luanoprzy** [comment]
I have the Same Problem (5.1)

**Comment by andikrueger** [comment]
Have you had the chance to test with the Graph Module V2.1.0?

**Comment by luanoprzy** [comment]
> Have you had the chance to test with the Graph Module V2.1.0?

Graph Module 2.1.0 is already released in Azure Automation.The same error message.

**Comment by Roladin** [comment]
@peombwa Hello again mate, unfortunately the issue persists with version 2.1.0

**Comment by RTTrickster** [comment]
Confirm I am also experiencing this issue

**Comment by peombwa** [comment]
> Could you try importing the module to Azure Automation again to see if the issue is still present? It does appear that there was an issue in Azure Automation that was preventing modules from being imported, which has since been resolved - [Azure/azure-powershell#21647 (comment)](https://github.com/Azure/azure-powershell/issues/21647#issuecomment-1632892165).
> 
> Please open a support ticket or question with the Azure Automation Account Team if the problem persists as the issue appears to be isolated to a few Automation Accounts (AA). See https://learn.microsoft.com/azure/automation/troubleshoot/automation-account. We are able to successfully import the module on new AA accounts.
> 

As stated above, the issue is external to the SDK. The stack trace shows the intermittent error occurring in Azure Automation during installation. We've notified the Azure Automation team of this error and they are currently investigating. Please open a ticket/question with Azure Automation as suggested above.

**Comment by DamirZelenic** [comment]
Has anyone noticed similar behavior with Azure Functions? I tried to upgrade to 2.0 and function starts failing with error:

_ERROR: The 'Connect-MgGraph' command was found in the module 'Microsoft.Graph.Authentication', but the module could not be loaded. For more information, run 'Import-Module Microsoft.Graph.Authentication'. Exception : Type : System.Management.Automation.CommandNotFoundException ErrorRecord : Exception : Type : System.Management.Automation.ParentContainsErrorRecordException Message : The 'Connect-MgGraph' command was found in the module 'Microsoft.Graph.Authentication', but the module could not be loaded. For more information, run 'Import-Module Microsoft.Graph.Authentication'. HResult : -2146233087 TargetObject : Connect-MgGraph CategoryInfo : ObjectNotFound: (Connect-MgGraph:String). ..._

Function is running 7.2 Core version. Azure modules are also used, updated to last version. If I rollback to 1.0 everything starts working again.

I am using requirements.psd1.
Working:

```
 'Microsoft.Graph.Authentication' = '1.*'
 'Microsoft.Graph.Users' = '1.*'
 'Microsoft.Graph.Applications' = '1.*'
 'Az' = '10.*'
 'AzTable' = '2.*'
```

Not working:
```
'Microsoft.Graph.Authentication' = '2.*'
'Microsoft.Graph.Users' = '2.*'
'Microsoft.Graph.Applications' = '2.*'
'Az' = '10.*'
'AzTable' = '2.*'
```

**Comment by LOldfield** [comment]
Issue persisting with powershell 5.1 and v2.1.0
Weirdly enough I'm getting a similar error trying to import the python msgraph-sdk wheel file to python packages: 
`Orchestrator.Activities.PythonPackageExtractException: Error while extractinig Python package: pip install failed for python package msgraph_sdk. at Orchestrator.Activities.SetModuleActivity.ExecuteInternal(CodeActivityContext context, Byte[] moduleContent, String moduleName, ModuleLanguage moduleLanguage, Guid moduleVersionId, String modulePath) at Orchestrator.Activities.SetModuleActivity.Execute(CodeActivityContext context) at System.Activities.CodeActivity.InternalExecute(ActivityInstance instance, ActivityExecutor executor, BookmarkManager bookmarkManager) at System.Activities.Runtime.ActivityExecutor.ExecuteActivityWorkItem.ExecuteBody(ActivityExecutor executor, BookmarkManager bookmarkManager, Location resultLocation)`

Other packages import without issue

**Comment by MrMKJ** [comment]
Same issue here... I successfully imported the Microsoft.Graph.Authentication module to my test Azure Automation account a couple of months ago, but when I tried to import it to the identical live account today it repeatedly failed with the same error as above.  (I'm targeting PS 5.1 as 7.1 and 7.2 are still listed as "Preview" in the module import dialogue)
 
_"Error importing the module Microsoft.Graph.Authentication. Import failed with the following error: Orchestrator.Shared.AsyncModuleImport.ModuleImportException: While importing the module, an error occurred while processing the module content. Internal error message: Exception has been thrown by the target of an invocation.. at Orchestrator.Activities.SetModuleVersion.ExecuteInternal(CodeActivityContext context, String moduleName, ModuleLanguage moduleLanguage, String pythonModuleVersion, String modulePath, Guid accountId, Guid moduleVersionId, Int64 moduleContentByteSize, String storageUri, Int32 moduleVersion) at Orchestrator.Activities.SetModuleVersion.Execute(CodeActivityContext context) at System.Activities.CodeActivity.InternalExecute(ActivityInstance instance, ActivityExecutor executor, BookmarkManager bookmarkManager) at System.Activities.Runtime.ActivityExecutor.ExecuteActivityWorkItem.ExecuteBody(ActivityExecutor executor, BookmarkManager bookmarkManager, Location resultLocation)"_

**Comment by RTTrickster** [comment]
As advised by @peombwa I've raised a support ticket with MS, who have acknowledged the issue and are investigating. They've indicated potentially it is only an issue in the Australia region, but I'm not sure if that's confirmed, or if it aligns with the other reports on this issue. 

If you are desperate - you can successfully still deploy v1.2.8 to Automation. Bear in mind that this module is a dependency for all other Graph modules, so you would have to use v1.2.8 of all those modules too.

**Comment by MrMKJ** [comment]
Hi @RTTrickster thanks for the update.  Re the affected region... defiantly not just Australia, I'm working in UK South.

**Comment by LucasHallGPE** [comment]
Also having the same issue in UK South.

**Comment by luanoprzy** [comment]
I'm working in Germany West Central

**Comment by A-J-A** [comment]
Also in UK South, I'm unable to open a support request as its not included with my subscription.
![image](https://github.com/microsoftgraph/msgraph-sdk-powershell/assets/32280491/b07ef8fc-1e98-4840-b41b-60c380dd43a0)

**Comment by anoopkv39** [comment]
> As advised by @peombwa I've raised a support ticket with MS, who have acknowledged the issue and are investigating. They've indicated potentially it is only an issue in the Australia region, but I'm not sure if that's confirmed, or if it aligns with the other reports on this issue.
> 
> If you are desperate - you can successfully still deploy v1.28 to Automation. Bear in mind that this module is a dependency for all other Graph modules, so you would have to use v1.28 of all those modules too.

@RTTrickster - did MS provide any update on this issue?

**Comment by RTTrickster** [comment]
> > As advised by @peombwa I've raised a support ticket with MS, who have acknowledged the issue and are investigating. They've indicated potentially it is only an issue in the Australia region, but I'm not sure if that's confirmed, or if it aligns with the other reports on this issue.
> > If you are desperate - you can successfully still deploy v1.28 to Automation. Bear in mind that this module is a dependency for all other Graph modules, so you would have to use v1.28 of all those modules too.
> 
> @RTTrickster - did MS provide any update on this issue?

They confirmed earlier today that it is definitely a multi-region issue and they are investigating.

**Comment by akvMS** [comment]
> > > As advised by @peombwa I've raised a support ticket with MS, who have acknowledged the issue and are investigating. They've indicated potentially it is only an issue in the Australia region, but I'm not sure if that's confirmed, or if it aligns with the other reports on this issue.
> > > If you are desperate - you can successfully still deploy v1.28 to Automation. Bear in mind that this module is a dependency for all other Graph modules, so you would have to use v1.28 of all those modules too.
> > 
> > 
> > @RTTrickster - did MS provide any update on this issue?
> 
> They confirmed earlier today that it is definitely a multi-region issue and they are investigating.

thanks. looks like it's a global issue.

**Comment by dvillartx** [comment]
Thank you ...Please keep us updated.

**Comment by dvillartx** [comment]
Any news on this?

**Comment by RTTrickster** [comment]
Update from MS on my support ticket; 

_Get update from Product Group. The ETA of end of August to fix to be deployed. At present, we may need to use version 1.28.0._

**Comment by Roladin** [comment]
thank you @RTTrickster

**Comment by davroscowley** [comment]
I am having the same issue when importing version 2.2.0, regardless of runtime. I have imported version 1.28.0 without any issue.

**Comment by kpinel** [comment]
> I am having the same issue when importing version 2.2.0, regardless of runtime. I have imported version 1.28.0 without any issue.

OK. Dumb question here.
How do you import older versions into Azure automation? Clearly I'm missing something in the process.

**Comment by akvMS** [comment]
> Update for those interested on my support ticket;
> 
> _We get update from Product Group on Tuesday, but they don’t say the date to fix the module issue. The issue is related to new SQL PAL changes and Product Group will revert the changes. Also, they suggest using version 1.28.0 at present which we already test it.
> 
> I will ask for product Group when we can import the new version module today. Once I get new update, I will email you._

@RTTrickster - did you hear any update form Microsoft/ Product Group?

**Comment by RTTrickster** [comment]
Last comm's from MS Support was that it would be fixed by end-Aug and to simply try again then. I have tested again today and still failing, so I have queried them for a status update. Will keep this forum posted on outcome.

**Comment by nguyenanhducs** [comment]
I also got the same issue when importing Microsoft.Graph module to Automation Account using Terraform. The error was: 
```
Orchestrator.Shared.AsyncModuleImport.ModuleImportException: An error occurred when trying to import the module to an internal PowerShell session. Either the module dependencies are not imported correctly or the module is unsupported. Internal Error Message: The required module 'Microsoft.Graph.Authentication' is not loaded. Load the module or remove the module from 'RequiredModules' in the file 'C:\Users\Client\Temp\WNKNZVUSSF\Microsoft.Graph\Microsoft.Graph.psd1'.

   at Orchestrator.Activities.SetModuleVersion.GetPsModuleVersion(String moduleName, String moduleFullPath, Guid accountId, Guid moduleVersionId, InitialSessionState defaultSessionState)
   at Orchestrator.Activities.SetModuleVersion.InsertModuleVersion(String moduleName, ModuleLanguage moduleLanguage, String pythonModuleVersion, String moduleFullyQualifiedName, ActivitiesMetadataWorkflowExtension activitiesMetadataWorkflowExtension, Guid accountId, Guid moduleVersionId, Int64 moduleContentByteSize, ContentUriInfo contentUri, String storageUri, Int32 moduleVersion, String runtimeEnvironment)
   at Orchestrator.Activities.SetModuleVersion.ExecuteInternal(CodeActivityContext context, String moduleName, ModuleLanguage moduleLanguage, String pythonModuleVersion, String modulePath, Guid accountId, Guid moduleVersionId, Int64 moduleContentByteSize, String storageUri, Int32 moduleVersion, String runtimeEnvironment)
   at Orchestrator.Activities.SetModuleVersion.Execute(CodeActivityContext context)
   at System.Activities.CodeActivity.InternalExecute(ActivityInstance instance, ActivityExecutor executor, BookmarkManager bookmarkManager)
   at System.Activities.Runtime.ActivityExecutor.ExecuteActivityWorkItem.ExecuteBody(ActivityExecutor executor, BookmarkManager bookmarkManager, Location resultLocation)
```

**Comment by NeitherSound** [comment]
Still failing with runtime 7.2

**Comment by PaoloCon** [comment]
I think it works now with module version 2.5.0 on runtime 5.1

<img width="668" alt="image" src="https://github.com/microsoftgraph/msgraph-sdk-powershell/assets/92427941/3dd3cbeb-69df-4f7a-8038-8e6d54c8053b">

Can someone else confirm?

**Comment by woter1832** [comment]
@PaoloCon,

I can confirm it does **not** import with runtime 5.1, but it does import with 7.2.

**Comment by akvMS** [comment]
> I think it works now with module version 2.5.0 on runtime 5.1
> 
> <img alt="image" width="668" src="https://user-images.githubusercontent.com/92427941/268587853-3dd3cbeb-69df-4f7a-8038-8e6d54c8053b.png">
> Can someone else confirm?

It did work manually for module version 2.5.0 on runtime 5.1. However, still failing with Terraform.

**Comment by kpinel** [comment]
This does work. I've updated my runbook modules from 1.28 to 2.6

Kind regards,
Kevin
-------------------------------
If you're going to be passionate about something, be passionate about
learning

**Comment by woter1832** [comment]
@kpinel, I'm not sure how or why it worked for you, but 11 hours ago, MSFT released version `2.6.1` which I am now able to import.

**Comment by wisewtf** [comment]
Same error today on a newly create AA.

```
Error importing the module Microsoft.Graph.Authentication. Import failed with the following error: Orchestrator.Shared.AsyncModuleImport.ModuleImportException: While importing the module, an error occurred while processing the module content. Internal error message: Exception has been thrown by the target of an invocation.. at Orchestrator.Activities.SetModuleVersion.ExecuteInternal(CodeActivityContext context, String moduleName, ModuleLanguage moduleLanguage, String pythonModuleVersion, String modulePath, Guid accountId, Guid moduleVersionId, Int64 moduleContentByteSize, String storageUri, Int32 moduleVersion) at Orchestrator.Activities.SetModuleVersion.Execute(CodeActivityContext context) at System.Activities.CodeActivity.InternalExecute(ActivityInstance instance, ActivityExecutor executor, BookmarkManager bookmarkManager) at System.Activities.Runtime.ActivityExecutor.ExecuteActivityWorkItem.ExecuteBody(ActivityExecutor executor, BookmarkManager bookmarkManager, Location resultLocation)
```
